### PR TITLE
Added support and test for custom artifact names in WandbLogger

### LINKED ISSUE #15990: WandbLogger add Artifact name selection and save by reference
### Description & Motivation

The model artifact is currently saved as:

```python
artifact = wandb.Artifact(name=f"model-{self.experiment.id}", type="model", metadata=metadata)
```
https://github.com/Lightning-AI/lightning/blob/1657ea8a39c3795e1276c491eff4f13b06a06e61/src/pytorch_lightning/loggers/wandb.py#L599

This results in obscure names such as: `model-3j2t0k7n` that is hard to refer back to the experiment (without looking at the lineages).

There is no current option to log the model by reference. So the model Artifact is always uploaded to WandB.
https://github.com/Lightning-AI/lightning/blob/1657ea8a39c3795e1276c491eff4f13b06a06e61/src/pytorch_lightning/loggers/wandb.py#L600 

### Pitch

For the naming issue, add the option to save the model by experiment name. This makes sense because there should be a 1-to-1 relationship between the experiment and the model. As long as experiment names are uniquely defined within a project, things should be okay.

```python
artifact = wandb.Artifact(name=f"model-{self.experiment.name}", type="model", metadata=metadata)
```

On log by reference point, add a flag to choose to log the Artifact by reference or not.

```python
artifact = wandb.Artifact(name=f"model-", type="model", metadata=metadata)

if log_by_reference:
    artifact.add_reference(p, name="model.ckpt")
else:
    artifact.add_file(p, name="model.ckpt")

self.experiment.log_artifact(artifact, aliases=[tag])
```

### Alternatives

_No response_

### Additional context

_No response_

cc @borda @awaelchli @morganmcg1 @borisdayma @scottire @manangoel99

### PR DESCRIPTION
**Pull request by manangoel99**
## Support for custom artifact names in WandbLogger

This PR adds a new optional argument to the WandbLogger for a custom artifact name for the model checkpoint. In case a name is not provided, the logger will continue with the original behaviour.

Fixes #15990 partially

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/pytorch_lightning/loggers/wandb.py [hunk 1]
```diff
@@ -359,6 +361,11 @@ def __init__(
             wandb.require("service")
             _ = self.experiment
 
+        if checkpoint_name:
+            self.checkpoint_name = checkpoint_name
+        else:
+            self.checkpoint_name = None
```
[the remark this answers is not in the document]
Comment (carmocca) [reply src/pytorch_lightning/loggers/wandb.py:367]: It's None by default
```suggestion
        self.checkpoint_name = checkpoint_name
```

File: src/pytorch_lightning/loggers/wandb.py [hunk 2]
```diff
@@ -303,6 +303,8 @@ def __init__(
         log_model: Union[str, bool] = False,
         experiment: Union[Run, RunDisabled, None] = None,
         prefix: str = "",
+        checkpoint_name: Optional[str] = None,
+        use_reference: bool = False,
```
Comment (carmocca) [inline src/pytorch_lightning/loggers/wandb.py:307]: How is this used?

Comment (manangoel99) [reply src/pytorch_lightning/loggers/wandb.py]: I've removed the `use_reference` argument

File: src/pytorch_lightning/loggers/wandb.py [hunk 3]
```diff
@@ -303,6 +303,7 @@ def __init__(
         log_model: Union[str, bool] = False,
         experiment: Union[Run, RunDisabled, None] = None,
         prefix: str = "",
+        checkpoint_name: Optional[str] = None,
```
Comment (awaelchli) [inline src/pytorch_lightning/loggers/wandb.py:306]: An update to the docstring (above) is missing

File: src/pytorch_lightning/loggers/wandb.py [hunk 4]
```diff
@@ -359,6 +360,8 @@ def __init__(
             wandb.require("service")
             _ = self.experiment
 
+        self.checkpoint_name = checkpoint_name
```
Comment (awaelchli) [inline src/pytorch_lightning/loggers/wandb.py:363]:
```suggestion
        self._checkpoint_name = checkpoint_name
```

**Review by awaelchli — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
LGTM after the pending comments are resolved.

**Pull request merged**